src/tools/rest-backend/authentication_application.cpp

```diff
@@ -0,0 +1,307 @@
+#include <ctime>
+#include <iostream>
+#include <regex>
+#include <stdlib.h>
+#include <string.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <jwt/jwt.h>
+#include <openssl/sha.h>
+
+#include "authentication_application.hpp"
+#include "crypto.hpp"
+#include "root_application.hpp"
+#include "service.hpp"
+#include "user_application.hpp"
+
+namespace kdbrest
+{
+
+/**
+ * @brief the constructor fot the authentication endopint application.
+ * @param srv a service container
+ */
+AuthenticationApp::AuthenticationApp (cppcms::service & srv) : cppcms::application (srv)
+{
+
+	dispatcher ().assign ("", &AuthenticationApp::authenticate, this);
+	mapper ().assign ("");
+}
+
+/**
+ * @brief handler for the authentication resource.
+ *		  attempts an authentication with the submitted credentials. in case the
+ *		  attempt succeeds, a session token will be returned, otherwise an error.
+ */
+void AuthenticationApp::authenticate ()
+{
+	RootApp::setCORSHeaders (response (), "POST,OPTIONS");
+
+	if (request ().request_method () == "POST")
+	{
+		// check if request data is of type application/json
+		if (request ().content_type_parsed ().media_type () != "application/json")
+		{
+			RootApp::setNotAcceptable (response (), "You have supplied an unsupported Content-Type.",
+						   "REQUEST_UNSUPPORTED_CONTENT_TYPE");
+			return;
+		}
+
+		// try to parse request body data
+		cppcms::json::value requestData;
+		try
+		{
+			requestData = RootApp::parsePostDataAsJson (request ());
+		}
+		catch (kdbrest::exception::InvalidPostDataFormatException & e)
+		{
+			RootApp::setBadRequest (response (), "The submitted data is not of type application/json.",
```

Comment:
At some places exceptions are abused for control flow. If you could avoid at least this it would be better. Your style is very exception-centric, which would be hard to change now. Hopefully they are not too expensive, did you benchmark it?